lib/Sema/CSApply.cpp

```diff
@@ -2831,7 +2831,23 @@ namespace {
     }
 
     Expr *visitTupleExpr(TupleExpr *expr) {
-      return simplifyExprType(expr);
+      auto toType = simplifyType(cs.getType(expr))->castTo<TupleType>();
```

Comment:
I think coerceToType() with a TupleExpr already coerces each element like this -- can you check?

Also, maybe it's the outer expression that should be coercing it's child, and not the TupleExpr itself that's wrong? What if you have a non-tuple expression there, does it work?

Reply:
Ah yes, I checked - it only supports tuple-to-tuple conversions in coerceToType. The problem is only related to tuples for binary operators and function calls, I'm going to add more test-cases which use function, prefix and postfix with single and two arguments to make sure everything works fine.

Reply:
coerceTupleToTuple() handles the case where the source is a TupleExpr and coerces the elements I think.

Reply:
I still think the problem is somebody is not coercing their argument (which might not be a TupleExpr).

Reply:
I will take a look at coerceToTupleType thank you! But I think it’s effectively a noop when types are the same. Also I can’t come up with example to reproduce crash with paren, because it simply forwards it’s sub-expr type to wrap into paren type, there can’t be a mismatch like in case of tuple.

Reply:
coerceToType() calls coerceTupleToTuple().

I still think this is the wrong fix. TupleExprs appear throughout the compiler in many places, and we've never had to do this until now. It only seems to happen in this one corner case that the type of the tuple expression does not match the types of its elements.

If you look at the rule in CSGen, the type of the TupleExpr is just a TupleType of the generated types of its elements. So imagine the elements are type variables $tv1 and $tv2, then the TupleExpr's type should be ($tv1, $tv2). I don't understand how cs.getType(tupleExpr) can return a TupleType whose element types don't match cs.getType() on the two sub-expressions.

cc @rudkx

Reply:
It could be that the ExprRewriter is producing the wrong-typed expression in a child element of the tuple. Can you do cs.getType() on the tuple, its elements ,and compare those with calling expr->getType() on the elements?

Reply:
I suggest you take a look at -debug-constraints output to understand this better, as I mentioned in my comment that tuple element would get a different type from what try? produces because it binds to AnyObject? (which comes from parameter-to-argument match) and checks if it’s convertable to P. It didn’t happen before (try it in -swift-version 4) because call would get P? assigned directly based on OptionalObject constraint and check id that’s equal to AnyObject?

Reply:
It sounds like `try?` behavior was changed recently in -swift-version 5? I think there's a bug in your code.

Also note that your test runs with -swift-version 4, not 5.

Reply:
Also note that CSGen would set types of elements to P directly but generate type variables for tuple type elements which are connected to element types via conversion constraints, this is the main source of mismatch here, because type of tuple would be different from its element types when type variables are bound.

Reply:
It was changed by SE-0230

Reply:
Since the tuple is an argument list, why isn't coerceCallArguments() doing the right thing here?

Reply:
Ah I will fix the test, I keep forgetting that 5 is not default yet...

Reply:
Note that argument conversions are common and well-tested. Why is this new implementation of `try?` the _only_ example that triggers this bad behavior?

Reply:
It’s not kicking in because parameter types are equal to what tuple type has for its elements, which is not used to be the case when we used OptinalBinding constraint.

Reply:
Only example so far maybe? I don’t have all of the answers unfortunately. I think we just generaly don’t generate disjunctions for argument conversion constraints (like in cases when IUO involved) that’s why it’s not that wide spred...

Reply:
https://github.com/apple/swift/pull/21348